### PR TITLE
Highlighter: Fix hinting state for keyboard-based selection

### PR DESCRIPTION
**Pull request by omnibs**
# :wrench: Modifying a component

## Context

TL;DR: Shift+arrow now only apply the highlight when shift is released. "Hinting" mode is used while shift is pressed.

The way Highlighter keyboard-based selection worked made it very difficult to implement annotation functionality for writing.

Successive focus messages after each shift+arrow would throw a wrench on our code, and immediate creation of highlights would force us create+expand+expand+expand highlights, which would generate backend messages on each cycle.

Making kb-based selection work like the mouse, using hinting first, then applying marks to hinted segments all at once, when shift is released, is an elegant solution that not only completely solves our problems at the writing annotations side, but also opens the door to add friendlier screenreader cues to hinting state.

**Specific things I want feedback on:** Is the hinting screenreader cue ok?

## :framed_picture: What does this change look like?

https://github.com/user-attachments/assets/db6fa025-26bf-4184-aeb8-8b3bf674d739

## Component completion checklist

- [x] I've gone through the relevant sections of the [Development Accessibility guide](https://paper.dropbox.com/doc/Accessibility-guide-4-Development--BiIVdijSaoijjOuhz3iTCJJ1Ag-rGoHpC91pFg3zTrYpvOCQ) with the changes I made to this component in mind
- [x] Changes are clearly documented
  - [x] Component docs include a changelog
  - [ x Any new exposed functions or properties have docs
- [x] Changes extend to the Component Catalog
  - [x] The Component Catalog is updated to use the newest version, if appropriate
  - [x] The Component Catalog example version number is updated, if appropriate
  - [x] Any new customizations are available from the Component Catalog
  - [x] The component example still has:
    - an accurate preview
    - valid sample code
    - correct keyboard behavior
    - correct and comprehensive guidance around how to use the component
- [x] Changes to the component are tested/the new version of the component is tested
- [x] Component API follows standard patterns in noredink-ui
  - e.g., as a dev, I can conveniently add an `nriDescription`
  - and adding a new feature to the component will _not_ require major API changes to the component
- [x] If this is a new major version of the component, our team has stories created to upgrade all instances of the old component. Here are links to the stories:
  - add your story links here OR just write this is not a new major version
- [x] Please assign the following reviewers:
  - [x] Someone from your team who can review your PR in full and review requirements from your team's perspective.
  - [x] Component library owner - Someone from this group will review your PR for accessibility and adherence to component library foundations.
  - [x] If there are user-facing changes, a designer. (You may want to direct your designer to the [deploy preview](https://github.com/NoRedInk/noredink-ui#reviews--preview-environments) for easy review):
    - For writing-related component changes, add Stacey (staceyadams)
    - For quiz engine-related components, add Ravi (ravi-morbia)
    - For a11y-related changes to general components, add Ben (bendansby)
    - For general component-related changes or if you’re not sure about something, add the Design group (NoRedInk/design)

### CODE REVIEW

**Review by brian-carroll — COMMENTED** [review]
Looks good! Couple of suggestions, nothing important. I'll leave it to Charbel or Ben to approve.

File: src/Nri/Ui/Highlighter/V5.elm [hunk 1]
```diff
@@ -1426,9 +1438,16 @@ viewHighlightable { renderMarkdown, overlaps } config highlightable =
                         , Key.shiftLeft (Keyboard <| SelectionExpandLeft highlightable.index)
                         ]
                     , Key.onKeyUpPreventDefault
-                        [ Key.shiftRight (Keyboard <| SelectionApplyTool highlightable.index)
-                        , Key.shiftLeft (Keyboard <| SelectionApplyTool highlightable.index)
-                        , Key.shift (Keyboard <| SelectionReset highlightable.index)
+                        [ -- Key.shift doesn't work on keyUp, bc `shiftKey` is False on keyUp
+                          { keyCode = 16
```
[the remark this answers is not in the document]
Comment (omnibs) [reply src/Nri/Ui/Highlighter/V5.elm:1442]: Oh good call on doing `shift.keyCode`. `{ shift | shiftKey = False}` would require a let assignment bc it's a function that takes a parameter. `shift.keyCode` is neater, then.

Yep 16 is the keycode for shift.

The `shiftKey` field indicates whether `shift` was held during the event, so you can disambiguate cases like this, where there's no unique keycode for a shifted right arrow, like there probably are for uppercase letters.

For the weird behavior that forces us to do this, I can guess w3c thought "when you release shift you're not holding shift amirite, so shiftKey gotta be `false` on keyup but `true` on keydown."

**Comment by omnibs** [comment]
@bendansby yes!

**Comment by omnibs** [comment]
on the numbering, the number is up to the use-case.

quizzes will use real names like "selecting evidence", writing is using weird numbers rn but we have an issue to fix it

**Comment by bendansby** [comment]
Technically speaking, one is not selecting a comment, but selecting text to make a comment. Selecting a comment would be a different action. But perhaps the intention is to abbreviate a complex phrase into something simple? What do you think?

**Comment by omnibs** [comment]
"Selecting text for [highlight-name]" is not terribly longer. Do you think it's clearer / less confusing?

**Comment by bendansby** [comment]
I do! If you agree, let's go with that.

File: src/Nri/Ui/Highlighter/V5.elm (hunk 1 above)
[the remark this answers is not in the document]
Comment (omnibs) [reply src/Nri/Ui/Highlighter/V5.elm]: can't `shift.keyCode` bc it needs a parameter, so i did a small compromise. pushed it!

File: src/Nri/Ui/Highlighter/V5.elm [hunk 2]
```diff
@@ -570,6 +572,16 @@ performAction action ( model, cmds ) =
             ( { model | selectionStartIndex = Nothing, selectionEndIndex = Nothing }, cmds )
 
 
+isFirstLastHinted : Maybe ( Int, Int ) -> Highlightable marker -> Bool
+isFirstLastHinted hintingIndices { index } =
```
[the remark this answers is not in the document]
Comment (omnibs) [reply src/Nri/Ui/Highlighter/V5.elm]: done!

**Comment by omnibs** [comment]
updated the video w/ the new msg, pushed addressing all review feedback so far =]

**Review by bendansby — APPROVED** [review]

**Review by brian-carroll — APPROVED** [review]

**Review by charbelrami — APPROVED** [review]
Looks great to me! Thanks for improving accessibility!
